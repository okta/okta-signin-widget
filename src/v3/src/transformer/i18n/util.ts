/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  getI18nKey,
  getI18NParams,
} from '../../../../v2/ion/i18nUtils';
import {
  FieldElement, TokenReplacement, UISchemaElement,
} from '../../types';
import { loc } from '../../util';

export const addTranslation = ({
  element,
  name,
  i18nKey,
  params,
  defaultValue = '',
  tokenReplacement,
  noTranslate,
}: {
  element: UISchemaElement;
  name: string;
  i18nKey: string;
  params?: any;
  defaultValue?: string;
  tokenReplacement?: TokenReplacement,
  noTranslate?: boolean,
}): void => {
  // TODO: change translations to required field with default value
  // eslint-disable-next-line no-param-reassign
  element.translations = element.translations || [];
  const useDefault = !i18nKey || (noTranslate && !!defaultValue);
  element.translations.push({
    name,
    i18nKey,
    value: useDefault ? defaultValue : loc(i18nKey, 'login', params, tokenReplacement),
    noTranslate,
  });
};

export const addLabelTranslationToFieldElement = (
  transaction: IdxTransaction,
  element: FieldElement,
): void => {
  const { nextStep } = transaction;
  const { relatesTo, name: stepName } = nextStep!;
  const authenticatorKey = relatesTo?.value?.key;
  const fieldName = element.options.inputMeta.name;
  const path = authenticatorKey
    ? `${stepName}.${authenticatorKey}.${fieldName}`
    : `${stepName}.${fieldName}`;
  const i18nKey = getI18nKey(path);
  const params = getI18NParams(nextStep, authenticatorKey);
  if (i18nKey || element.label) {
    const noTranslate = element.options.inputMeta.customLabel;
    addTranslation({
      element, name: 'label', i18nKey, params, defaultValue: element.label, noTranslate,
    });
  }
};

export const addOptionalLabelTranslationToFieldElement = (
  element: FieldElement,
): void => {
  if (element.options.inputMeta.required === false) {
    addTranslation({
      element, name: 'optionalLabel', i18nKey: 'oie.form.field.optional',
    });
  }
};
