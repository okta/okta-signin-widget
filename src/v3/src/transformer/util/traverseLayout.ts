/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FormBag,
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: FormBag['uischema'];
  predicate: PredicateFn;
  callback: CallbackFn;
};

export const traverseLayout = (options: Options) => {
  const fn = (
    layout: FormBag['uischema'],
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
};
