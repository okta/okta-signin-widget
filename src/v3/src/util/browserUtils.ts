/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const isAndroid = (): boolean => (
  // Windows Phone also contains "Android"
  /android/i.test(navigator.userAgent)
    && !/windows phone/i.test(navigator.userAgent)
);

export const isIOS = (): boolean => (
  // iOS detection from: http://stackoverflow.com/a/9039885/177710
  /iPad|iPhone|iPod/i.test(navigator.userAgent)
    // iPad on iOS 13 detection
    || (navigator.userAgent.includes('Mac') && 'ontouchend' in document)
);

export const isAndroidOrIOS = (): boolean => isAndroid() || isIOS();
