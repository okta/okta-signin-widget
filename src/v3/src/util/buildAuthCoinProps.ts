/*
 * Copyright (c) 2025-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  AUTHENTICATOR_KEY, DEVICE_ENROLLMENT_TYPE, EMAIL_AUTHENTICATOR_TERMINAL_KEYS, IDX_STEP,
} from '../constants';
import { AuthCoinProps } from '../types';
import { getAuthenticatorKey } from './getAuthenticatorKey';
import { containsOneOfMessageKeys } from './idx';

export const buildAuthCoinProps = (
  transaction?: IdxTransaction | null,
): AuthCoinProps | undefined => {
  if (!transaction) {
    return undefined;
  }

  const { nextStep, messages } = transaction;
  if (containsOneOfMessageKeys(EMAIL_AUTHENTICATOR_TERMINAL_KEYS, messages)
    || nextStep?.name === IDX_STEP.CONSENT_EMAIL_CHALLENGE) {
    return { authenticatorKey: AUTHENTICATOR_KEY.EMAIL };
  }

  if (nextStep?.name === IDX_STEP.PIV_IDP) {
    return { authenticatorKey: IDX_STEP.PIV_IDP };
  }

  if (nextStep?.name === IDX_STEP.REDIRECT_IDVERIFY && nextStep?.idp?.id) {
    return { authenticatorKey: nextStep.idp.id };
  }

  // @ts-expect-error Property 'deviceEnrollment' does not exist on type 'IdxContext' ts(2339)
  if (transaction.context?.deviceEnrollment?.value?.name === DEVICE_ENROLLMENT_TYPE.ODA) {
    return { authenticatorKey: AUTHENTICATOR_KEY.OV };
  }

  const authenticatorKey = getAuthenticatorKey(transaction);
  if (authenticatorKey) {
    return {
      authenticatorKey,
      // @ts-ignore logoUri missing from interface
      url: nextStep?.relatesTo?.value.logoUri,
    };
  }

  return undefined;
};
