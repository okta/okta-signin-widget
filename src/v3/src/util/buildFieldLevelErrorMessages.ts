/*
 * Copyright (c) 2024-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { WidgetMessage } from '../types';

type FieldLevelErrorMessages = {
  errorMessage?: string;
  errorMessageList?: string[];
};

export const buildFieldLevelErrorMessages = (
  errors: WidgetMessage[] | undefined,
) : FieldLevelErrorMessages => {
  const hasErrors = typeof errors !== 'undefined' && errors.length > 0;

  if (hasErrors) {
    if (errors.length === 1) {
      const error = errors[0];
      // Covers nested messages, e.g. password requirements not met
      // The SIW does not have messages nested more than once
      if (Array.isArray(error.message)) {
        return {
          errorMessage: error.description,
          errorMessageList: error.message
            .filter((msg) => typeof msg.message === 'string')
            .map((msg) => msg.message as string),
        };
      }
      return { errorMessage: error.message };
    }
    return {
      errorMessageList: errors
        .filter((error) => typeof error.message === 'string')
        .map((error) => error.message as string),
    };
  }

  return {};
};
