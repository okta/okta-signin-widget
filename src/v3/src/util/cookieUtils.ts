/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import Cookies from 'js-cookie';

const DEFAULT_EXPIRY_DAYS = 365;
const DEFAULT_COOKIE_PATH = '/';
const LAST_USERNAME_COOKIE_NAME = 'ln';
const USER_AUTHENTICATED_COOKIE_NAME = 'isuauth';

export const setCookieString = (
  name: string,
  value: string,
  expiryDays?: number,
  path?: string,
): void => {
  Cookies.set(
    name,
    value,
    { expires: expiryDays ?? DEFAULT_EXPIRY_DAYS, path: path ?? DEFAULT_COOKIE_PATH },
  );
};

export const getCookieString = (name: string): string => Cookies.get(name) as string;

export const getCookieUserAuthenticated = (): string => getCookieString(
  USER_AUTHENTICATED_COOKIE_NAME,
);

export const removeCookie = (name: string): void => Cookies.remove(name);

export const setUsernameCookie = (username: string): void => setCookieString(
  LAST_USERNAME_COOKIE_NAME,
  username,
);

export const setCookieUserAuthenticated = (): void => {
  Cookies.set(
    USER_AUTHENTICATED_COOKIE_NAME,
    '1',
    { expires: 1, path: DEFAULT_COOKIE_PATH },
  );
};

export const getUsernameCookie = (): string => getCookieString(LAST_USERNAME_COOKIE_NAME);

export const removeUsernameCookie = (): void => removeCookie(LAST_USERNAME_COOKIE_NAME);

export const removeUserAuthenticatedCookie = (): void => removeCookie(
  USER_AUTHENTICATED_COOKIE_NAME,
);
