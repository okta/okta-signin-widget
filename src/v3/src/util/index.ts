/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export * from './browserUtils';
export * from './buildErrorMessageIds';
export * from './clipboard';
export * from './cookieUtils';
export * from './environmentUtils';
export * from './flattenInputs';
export * from './formatError';
export * from './formUtils';
export * from './generateRandomString';
export * from './getAuthenticatorKey';
export * from './getCurrentAuthenticator';
export * from './getElementKey';
export * from './getImmutableData';
export * from './getTranslation';
export * from './getValidationMessages';
export * from './idxUtils';
export * from './isInteractiveElement';
export * from './isPasswordRecovery';
export * from './isPollingStep';
export * from './languageUtils';
export * from './locUtil';
export * from './passwordUtils';
export * from './removeFieldLevelMessages';
export * from './resetMessagesToInputs';
export * from './settingsUtils';
export * from './shouldShowCancelLink';
export * from './toNestedObject';
export * from './webauthnUtils';
