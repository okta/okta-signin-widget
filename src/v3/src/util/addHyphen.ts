/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
*/
import { KeyboardEvent } from "src/types"

export const addHyphen = (e: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
    const currentVal = e.currentTarget.value;
    // Add hyphen after 4th character. includes() check allows deletion of hyphen and prevents double hyphens
    if (currentVal && currentVal.length === 4 && !['Backspace', 'Delete', '-'].includes(e.key)) {
      e.currentTarget.value = currentVal.concat('-');
    } else if (currentVal && currentVal.length > 4 && currentVal.charAt(4) !== '-') {
      // If user types too fast, the above concatenation can be skipped. This retroactively adds a hyphen if it wasn't added
      e.currentTarget.value = currentVal.slice(0, 4) + '-' + currentVal.slice(4);
    }
}
