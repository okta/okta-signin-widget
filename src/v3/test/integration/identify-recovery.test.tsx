/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { waitFor } from '@testing-library/preact';
import { createAuthJsPayloadArgs, setup } from './util';

import mockResponse from '../../src/mocks/response/idp/idx/recover/default.json';

describe('identify-recovery', () => {
  it('renders form', async () => {
    const { container, findByLabelText } = await setup({ mockResponse });
    await findByLabelText(/Email or Username/);
    expect(container).toMatchSnapshot();
  });

  it('fails client validation when missing identifier', async () => {
    const {
      user,
      findByLabelText,
      findByText,
      findAllByRole,
    } = await setup({ mockResponse });

    await user.click(await findByText('Next', { selector: 'button' }));
    const [globalError] = await findAllByRole('alert');
    expect(globalError.innerHTML).toContain('We found some errors. Please review the form and make corrections.');
    const identifierEl = await findByLabelText('Email or Username') as HTMLInputElement;
    expect(identifierEl).toHaveErrorMessage(/This field cannot be left blank$/);
  });

  it('sends correct payload', async () => {
    const {
      authClient,
      user,
      findByLabelText,
      findByText,
    } = await setup({ mockResponse });

    const titleElement = await findByText('Reset your password');
    await waitFor(() => expect(titleElement).toHaveFocus());
    const usernameEl = await findByLabelText('Email or Username') as HTMLInputElement;
    await user.type(usernameEl, 'testuser@okta.com');
    await user.click(await findByText('Next', { selector: 'button' }));

    expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
      ...createAuthJsPayloadArgs('POST', 'idp/idx/identify', { identifier: 'testuser@okta.com' }),
    );
  });
});
