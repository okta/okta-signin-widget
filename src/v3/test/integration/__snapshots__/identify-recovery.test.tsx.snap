// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`identify-recovery renders form 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <span>
      <main
        class="auth-container main-container mainViewContainer MuiBox-root emotion-1"
        data-commit="b9bbc0140703c3fbf0e2e58920362e70"
        data-version="0.0.0"
        id="okta-sign-in"
      >
        <div
          class="siwContainer MuiBox-root emotion-2"
        >
          <div
            class="okta-sign-in-header auth-header siwHeader"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 emotion-3"
            />
          </div>
          <div
            class="MuiBox-root emotion-4"
          >
            <form
              class="o-form"
              data-se="form"
              novalidate=""
            >
              <div
                class="MuiBox-root emotion-5"
              >
                <div
                  class="MuiBox-root emotion-6"
                >
                  <div
                    class="MuiBox-root emotion-7"
                  >
                    <h2
                      class="MuiTypography-root MuiTypography-h3 emotion-8"
                      data-se="o-form-head"
                    >
                      Reset your password
                    </h2>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <div
                    class="MuiBox-root emotion-10"
                  >
                    <label
                      class="MuiInputLabel-root MuiFormLabel-root MuiFormLabel-colorPrimary emotion-11"
                      for="identifier"
                    >
                      Email or Username
                    </label>
                    <div
                      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth emotion-12"
                    >
                      <input
                        aria-invalid="false"
                        autocomplete="username"
                        class="MuiOutlinedInput-input MuiInputBase-input emotion-13"
                        data-se="identifier"
                        id="identifier"
                        name="identifier"
                        type="text"
                      />
                      <fieldset
                        aria-hidden="true"
                        class="MuiOutlinedInput-notchedOutline emotion-14"
                      >
                        <legend
                          class="emotion-15"
                        >
                          <span
                            class="notranslate"
                          >
                            â€‹
                          </span>
                        </legend>
                      </fieldset>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <button
                    class="MuiButton-root MuiButton-primary MuiButton-primaryPrimary MuiButton-sizeMedium MuiButton-primarySizeMedium MuiButton-disableElevation MuiButton-fullWidth MuiButtonBase-root  emotion-17"
                    data-type="save"
                    tabindex="0"
                    type="submit"
                  >
                    Next
                  </button>
                </div>
                <div
                  class="MuiBox-root emotion-6"
                >
                  <a
                    class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways emotion-19"
                    data-se="cancel"
                    href="javascript:void(0)"
                  >
                    Back to sign in
                  </a>
                </div>
              </div>
            </form>
          </div>
        </div>
      </main>
    </span>
  </div>
</div>
`;
